packages/dds/legacy-dds/src/test/array/fuzzUtils.ts

```diff
@@ -0,0 +1,233 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "node:assert";
+
+import {
+	type Reducer,
+	combineReducers,
+	createWeightedGenerator,
+	takeAsync,
+} from "@fluid-private/stochastic-test-utils";
+import type { DDSFuzzModel, DDSFuzzTestState } from "@fluid-private/test-dds-utils";
+import type { Serializable } from "@fluidframework/datastore-definitions/internal";
+
+import type { SerializableTypeForSharedArray } from "../../index.js";
+import { SharedArrayFactory } from "../../index.js";
+
+/**
+ * Type for the SharedArray operation
+ *
+ * @internal
+ */
+export interface SharedArrayInsert<T> {
+	type: "insert";
+	index: number;
+	value: T;
+}
+
+/**
+ * Type for the SharedArray operation
+ *
+ * @internal
+ */
+export interface SharedArrayDelete {
+	type: "delete";
+	index: number;
+}
+
+/**
+ * Type for the SharedArray operation
+ *
+ * @internal
+ */
+export interface SharedArrayMove {
+	type: "move";
+	oldIndex: number;
+	newIndex: number;
+}
+
+/**
+ * Type for the SharedArray operation
+ *
+ * @internal
+ */
+export interface SharedArrayToggle {
+	type: "toggle";
+	entryId: string;
+}
+
+/**
+ * Type for the SharedArray operation
+ *
+ * @internal
+ */
+export interface SharedArrayToggleMove {
+	type: "toggleMove";
+	oldEntryId: string;
+	newEntryId: string;
+}
+
+/**
+ * Type for the SharedArray operation
+ *
+ * @internal
+ */
+export interface SharedArrayInsertBulkAfter<T> {
+	type: "insertBulkAfter";
+	ref: T | undefined;
+	values: T[];
+}
+
+/**
+ * SharedArray operations union type.
+ *
+ * @internal
+ */
+export type SharedArrayOperation<T> =
+	| SharedArrayInsert<T>
+	| SharedArrayDelete
+	| SharedArrayMove
+	| SharedArrayToggle
+	| SharedArrayToggleMove
+	| SharedArrayInsertBulkAfter<T>;
+
+/**
+ * Creates a reducer for SharedArray operations.
+ * @internal
+ */
+export function makeSharedArrayReducer<T extends SerializableTypeForSharedArray>(): Reducer<
+	SharedArrayOperation<T>,
+	DDSFuzzTestState<SharedArrayFactory<T>>
+> {
+	return combineReducers({
+		insert: ({ client }, { index, value }) => {
+			client.channel.insert(index, value as Serializable<typeof value> & T);
+		},
+		insertBulkAfter: ({ client }, { ref, values }) => {
+			client.channel.insertBulkAfter(
+				ref,
+				values.map((v) => v as Serializable<T> & T),
+			);
+		},
+		delete: ({ client }, { index }) => {
+			client.channel.delete(index);
+		},
+		move: ({ client }, { oldIndex, newIndex }) => {
+			client.channel.move(oldIndex, newIndex);
+		},
+		toggle: ({ client }, { entryId }) => {
+			client.channel.toggle(entryId);
+		},
+		toggleMove: ({ client }, { oldEntryId, newEntryId }) => {
+			client.channel.toggleMove(oldEntryId, newEntryId);
+		},
+	});
+}
+
+/**
+ * Creates a generator that yields SharedArray operations.
+ *
+ * @internal
+ */
+export function makeSharedArrayOperationGenerator<T extends SerializableTypeForSharedArray>(
```

Comment:
i don't think the type really matters, so i'd probably just hard code it to something, like string or number. then you can also get rid of the valuesPool, and just generate random data. that will reduce the likelihood of the state just happening to match because the values are so limited.